### ISSUE TITLE
insertText and FreetextAnnot doesnt work as expected with existing pdf

### ISSUE THREAD

**Issue by return71** [rk700/PyMuPDF#247]
Hello,
sorry for another post but eventually you can help me.
What im trying to do is automaticly comment electronic circuit diagrams in pdf format. 
Im search for a specific string in the pdf with `page.getTextWords()` . Then i want to add
to this position a text or a FreetextAnnotation.
The script runs fine, but the result is not what iam expect.
Please have a look at it. Here my script:

```
# -*- coding: utf-8 -*-

import fitz
import re, string

#only accept this chars because some pdf contains strange chars
legal_chars = string.ascii_lowercase + string.ascii_uppercase + string.digits + ",;:@!#$%&'*+-.^_`|~:"
p = '[' + re.escape(legal_chars) + ']' + '+'
p = re.compile(p)


    
doc = fitz.open("kicad.pdf")
page = doc[0]


fontsize = 6
color = (1, 0, 0)
text_to_find = "R12"
annotation = "F146"


x0 = 0
y0 = 0
for words in page.getTextWords():
    word = words[4]
    m = p.match(word)
    if m:
        if m.group() == text_to_find:
            x0 = words[0]
            y0 = words[1]
print("coordinates of R12", x0, y0)
page.insertText(fitz.Point(x0 , y0), annotation, fontsize = fontsize, color = color)
page.insertText(fitz.Point(362,217), "Test", 54)
annot = page.addFreetextAnnot(fitz.Rect(x0, y0, x0 + 100, y0 -100), "xxxx")
doc.save("test.pdf")
```

I will add some screenshots and the original pdf.
![addtext wrong position and size](https://user-images.githubusercontent.com/26299072/50935669-10f51a00-146d-11e9-8af8-2df4dce9f4c8.png)

![freetextannotation wrong position](https://user-images.githubusercontent.com/26299072/50935793-882aae00-146d-11e9-98a2-7d735ee26089.png)

[kicad.pdf](https://github.com/rk700/PyMuPDF/files/2742978/kicad.pdf)

**Comment by JorjMcKie** [comment]
Hi,
found the reason for the issue you are having.
For a new user of PyMuPDF your PDF is an unfortunate choice:

The page definition "source code" (the stream of the `/Contents` object) is uncleanly coded - it defines its own coordinate system, so all direct text additions of fitz will land in Nirwana.
"Clean" coding would be to encapsulate the source in a pair of strings "q" - "Q" (push to stack, pop from stack).

Therefore, I cleaned the page with low-level method `page._cleanContent()` before inserting my comment(s). Your (simplified) script now looks like this:

```python
# -*- coding: utf-8 -*-

import fitz
import re, string

#only accept this chars because some pdf contains strange chars
legal_chars = string.ascii_lowercase + string.ascii_uppercase + string.digits + ",;:@!#$%&'*+-.^_`|~:"
p = '[' + re.escape(legal_chars) + ']' + '+'
p = re.compile(p)

doc = fitz.open("kicad.pdf")
page = doc[0]
page._cleanContents() # repair page definition source

fontsize = 5.5
color = (1, 0, 0)
text_to_find = "R12"
annotation = " ** F146" # make comment more eye-catching

for words in page.getTextWords():
    if words[4] != text_to_find:
        continue
    r = fitz.Rect(words[:4])    # rect of the found word
    p = r.br                    # bottom-right = comment insertion point
    page.insertText(p,          # insertion point
                    annotation, # inserted comment
                    fontsize=fontsize, color=color)

doc.save("test.pdf")
```

It delivers this:

![marked](https://user-images.githubusercontent.com/8290722/50964732-6dc0f680-14a6-11e9-8745-99328e7b0c6f.JPG)

`page._cleanContents()` actually does a lot more cleaning work than we require here. But it should do its job in your case.

**Comment by return71** [comment]
Hello,
thank you for your help, `insertText `works now fine. I am happy with that.
The output from  `page.getText("dict")` looks a bit complicated to me, so i will stay with `page.getTextWords()`. 
Again, thank you. Your package helps me from doing a lot of manual text insertions. Sometimes i have
do annote several hundret notes in the schematics. Now i can make a little tool for this.

**Comment by JorjMcKie** [comment]
Great to hear that.
Do not hesitate to come back with more questions.

**Comment by return71** [comment]
Hello,
sorry for this post, but i have a lot of pdf's that are signed and only allow comments.
Doesnt know it befor i get to work. How difficult is it do get the FreeText annotations working.

**Comment by JorjMcKie** [comment]
FreeText does work. You have just do experiment a little until you reach a satisfactory appearance - fontsize, no border, etc. Just go ahead and try it.
Over the next few hours I cannot help because I have a major hardware crash due to electricity failures and subsequent high voltages ... :-(((

**Comment by JorjMcKie** [comment]
I am back. Damage under control.
Here is a script that looks for text "links" on all pages of `test.pdf` and
1. Surrounds the found text with a blue annotation rectangle
2. Adds a FreeText annotation in another rectangle **to the right** containing " <=" in red.
[Desktop.zip](https://github.com/rk700/PyMuPDF/files/2751487/Desktop.zip)

It also demonstrates how to control the appearance of the FreeText annot to some extent: no border, text color. I cannot prevent unfortunately that background is white ...

Hope it helps

**Comment by JorjMcKie** [comment]
Please do not forget to always `update` every annot if you made changes to the standard appearance after its creation. One update is sufficient after any number of changes.

**Comment by JorjMcKie** [comment]
Here is another version of the script. It uses `page.searchFor` to find the text. It may be even a bit simpler, because it delivers the found locations directly as rectangles (or - *new* - as quads). It also ignores upper / lower case.
I have included an effort to calculate the appropriate fontsize from the rectangle of the found text location (could have been done in the other example as well):
```python
import fitz
import sys, os
mydir = sys.path[0]
fname=os.path.join(mydir, "test.pdf")
doc=fitz.open(fname)
ofile=os.path.join(mydir, "test-commented.pdf")
blue = (0,0,1)
# similar to the other version, but uses 'searchFor' to
# find the text. Delivers directly the rectangles and ignores
# upp / lower case
for page in doc:
    rlist = page.searchFor("links", hit_max=20)
    for r in rlist:
        nr = r + (r.width, 0, r.width, 0) # = same size rect to the right
        a = page.addFreetextAnnot(nr, " <=")
        a.setBorder({"width":0})
        # try to calculate the fontsize:
        fsize = r.height / 1.2
        a.update(fontsize=fsize, text_color=blue)
        b = page.addRectAnnot(r)
        b.setBorder({"width":0.3})
        b.setColors({"stroke":blue})
        b.update()

doc.save(ofile, garbage=3, deflate=True)
```

**Comment by JorjMcKie** [comment]
In one of the next PyMuPDF versions, I will create 
* the option to have colorless FreeText backgrounds
* a way to calculate the text length based on chosen font and fontsize.

This will allow to choose adequate rectangle sizes that contain text without generated line breaks.

**Comment by return71** [comment]
Hello,
> I am back. Damage under control.

I am glad to hear that.

I am tried your `Desktop.zip`. But i will try to show you what i mean when i say FreeText Annotations
dont work as expected. For this i take your generated file `test-commented.pdf` and opened it in 
Acrobat Reader DC.
Then i deleted one FreeText Annotation and added a FreeText Annotation with Acrobat Reader DC.
Then i take some screenshots to show you the difference. I know the from me added annotation
has not exactly the same properties. 
But please look at the bounding rectangles and the look at different zooms. 
Here a screen with zoom 100%. Everything looks fine.
![zoom100](https://user-images.githubusercontent.com/26299072/51071312-61c46880-164f-11e9-946f-4839c5d34511.jpg)

Here a screen with zoom 400%. Things looks now different.
![zoom400](https://user-images.githubusercontent.com/26299072/51071331-8c162600-164f-11e9-9fa6-c984347d0d22.jpg)

Sorry, i will show two more screens with properties to show the difference in the bounding box
of the annotation.
Here a screen from the annotation added with pymupdf
![pymupdf_zoom400_properties](https://user-images.githubusercontent.com/26299072/51071361-09419b00-1650-11e9-80e0-43004d1ba5b7.jpg)

Here a screen from the annotation added with Acrobat Reader DC
![adobereader_zoom400_properties](https://user-images.githubusercontent.com/26299072/51071381-4ad24600-1650-11e9-819c-16428fe4afb1.jpg)

The cause fore the different looks must or could be a bug in the underlying mupdf library. What to
you mean?

> In one of the next PyMuPDF versions, I will create
> 
>     the option to have colorless FreeText backgrounds
>     a way to calculate the text length based on chosen font and fontsize.
> 
> This will allow to choose adequate rectangle sizes that contain text without generated line breaks.

This would be a great improvment. But i find right at the time now, your package is the easiest package for pdf manipulation what i found on the internet. Even without your improvement.

**Comment by JorjMcKie** [comment]
Hi,
> bug in the underlying mupdf

maybe, but I am not so sure. I use about 8 different PDF viewers and most of them **agree**:
1. **MuPDF** itself (of course), **SumatraPDF**, Internet browsers **Edge** and **Firefox**, **PDF Xchange Reader** and **Nitro PDF** think that **MuPDF is right**
2. Only **Acrobat DC** and **Foxit Reader** **disagree**

All 8 agree about the FreeText rectangle, but Acrobat and Foxit have a different opinion about the font's size. Whatever, I will look into this later.

If you want, you can make a quick fix yourself, so that the box of FreeText Annots is no longer white, but becomes transparent. This part of the package is in Python language and can thus easily be changed:
Edit `fitz.py`, go to line 3581:
![grafik](https://user-images.githubusercontent.com/8290722/51071841-eacaa800-162d-11e9-8919-c5471781d02b.png)

and insert the following 3 statements:

![grafik](https://user-images.githubusercontent.com/8290722/51071859-5f054b80-162e-11e9-94d0-ea4fe766bf11.png)

This change will suppress the FreeText rectangle be drawn totally. Boder color and border width will also be ignored then.
This is the change I will publish with the next version - so no worries ;-).

**Comment by JorjMcKie** [comment]
Good you liked the hotfix.
But, as you noticed, this **stays** transparent only as long as you do not touch it with another annotation editor. Let me mention some background why this is so:

* the look (appearance) of an annotation is completely under control of the respective annotation editor (MuPDF, Foxit, Adobe, etc.). Every one of them has its own logic how things should be done.
* the appearance is defined in a special source code object, the "appearance stream". This object is connected with the annotation object itself. Every editor defines what it likes in this source code.
* if you edit an annotation, the editor **recreates** the appearance stream! The outcome is completely different between editors - even if the resulting look appears equal.

Example:
* The appearance stream after PyMuPDF insertion:

```
stream
q
BT
0 0 1 rg
13.799999 TL
/Helv 16.488 Tf
2 14.488037 Td
T*
(<==) Tj
ET
Q

endstream
```
* appearance stream after editing with Adobe:
```
stream
1 G
278.5999 657.1118 47 15.488 re
S
q
1 0 0 1 0 0 cm
280.0999 658.6118 44 12.488 re
W
n
0 g
0 G
BT
/Helv 12 Tf
0 g
280.1 661.625 Td
(<==) Tj
ET
Q
endstream
```
With Foxit:
```
q
1 1 1 rg
0 -14.4881 46 14.4881 re f
Q
q
0 0 0 RG
1 w
-0.5 -14.9881 47 15.4881 re S
Q
q
/Tx BMC
0 -14.4881 46 14.4881 re
W
n
BT
/FXF2 12 Tf
 0.000000 0.000000 1.000000 rg
0 -10.86 Td
(<==) Tj
ET
EMC
Q
endstream
```

And so on.

So, if you want to use (Py-) MuPDF then you must live with whatever it creates.
BTW: both, Adobe and Foxit again create different looking things if you use them for doing apparently the same result ...

The only thing I can additionally offer in the next version (apart from transparent text boxes) is a function determining the text length for a given font and fontsize. This will help sizing the box for FreeText correctly - you can use the text length as the (minimum) width for the FreeText rectangle.

**Comment by return71** [comment]
Hello,

> Let me mention some background why this is so:

Thanks for sharing some background. I didnt know what going on behind the scenes. Things are know much clearer to me. 

> So, if you want to use (Py-) MuPDF then you must live with whatever it creates.

Yes, i agree. Better then annotate hundrets or thousends measuring points by hand. ;-) 

> The only thing I can additionally offer...

As i said bevor, i like your packgage and such a function makes this even good packgage even better.

**Comment by JorjMcKie** [comment]
hello again,
just published a new version 1.14.6.
As promised, the length of a textstring can now be calculated for any fontsize and a built-in font. Modified example script:

```python
import fitz
"""
Demo script
-----------
Search for some text and insert a comment aside each occurrence
as a FreeText annotation.
"""
doc = fitz.open(fname)
ofile = "test-" + fname
blue = (0,0,1)
gold = (1,1,0)
search_text = "something"
comment = "*** comment ***"

# calculate comment length for font Helvetica and fontsize 1
c_len1 = fitz.getTextlen(comment, fontname="helv", fontsize=1)

# Use 'searchFor' to find the text. Delivers directly the rectangles
# and ignores upper / lower case.
for page in doc:
    rlist = page.searchFor(search_text, hit_max=20)

    for r in rlist:
        # calculate the fontsize as 5/6 of rect height
        fsize = r.height / 1.2

        # make comment rectangle to the right of found text. Its width will
        # be adjusted with the fontsize
        crect = fitz.Rect(r.x1, r.y0, r.x1 + c_len1 * fsize, r.y1)

        # add the comment as FreeText annotation
        a = page.addFreetextAnnot(crect, comment)
        a.setBorder({"width":0})            # not necessary: no fill color!

        # update annot with some properties
        a.update(fontsize=fsize, text_color=blue, fill_color=None)

        # for demonstration only: also enclose found text with a rectangle
        b = page.addRectAnnot(r)
        b.setBorder({"width":0.3})
        b.setColors({"stroke":blue})
        b.update()

doc.save(ofile, garbage=3, deflate=True)
```
